File: matrix_algorithms/src/matrices/Algoritmo12_IV4_ParallelBlock.java
package matrices;

import java.math.BigInteger;
import java.util.concurrent.*;

/**
 * Algoritmo 12: IV.4 Parallel Block
 * Tipo: Iterativo por bloques con paralelismo - orden i-k-j
 * Descripcion: Version paralela de IV.3 Sequential Block.
 * Usa el orden i-k-j dentro del bloque y distribuye bloques de filas
 * entre hilos. Combina el reuso de registros de IV.3 con el paralelismo.
 * Complejidad: O(n^3 / p)
 */
public class Algoritmo12_IV4_ParallelBlock {

    static final int BLOQUE = 64;
    static final int HILOS  = Runtime.getRuntime().availableProcessors();

    public static BigInteger[][] multiplicar(BigInteger[][] A, BigInteger[][] B) throws Exception {
        int n = A.length;
        BigInteger[][] C = new BigInteger[n][n];
        for (int i=0;i<n;i++) for (int j=0;j<n;j++) C[i][j]=BigInteger.ZERO;

        ExecutorService pool = Executors.newFixedThreadPool(HILOS);
        java.util.List<Future<?>> tareas = new java.util.ArrayList<>();

        for (int ii=0; ii<n; ii+=BLOQUE) {
            final int iiF = ii;
            tareas.add(pool.submit(() -> {
                int iM = Math.min(iiF+BLOQUE, n);
                for (int kk=0; kk<n; kk+=BLOQUE)
                    for (int jj=0; jj<n; jj+=BLOQUE) {
                        int kM=Math.min(kk+BLOQUE,n), jM=Math.min(jj+BLOQUE,n);
                        for (int i=iiF; i<iM; i++)
                            for (int k=kk; k<kM; k++) {
                                BigInteger aik = A[i][k];
                                for (int j=jj; j<jM; j++)
                                    C[i][j] = C[i][j].add(aik.multiply(B[k][j]));
                            }
                    }
            }));
        }
        pool.shutdown();
        for (Future<?> f : tareas) f.get();
        return C;
    }
}

File: matrix_algorithms_python/algoritmo12_IV4_parallel_block.py
# Algoritmo 12: IV.4 Parallel Block
# Complejidad: O(n^3 / p)
# Version paralela de IV.3 Sequential Block
# Usa orden i-k-j dentro del bloque y distribuye bloques de filas entre hilos

import threading

BLOQUE = 64
HILOS  = __import__('os').cpu_count()

def multiplicar(A, B):
    n = len(A)
    C = [[0] * n for _ in range(n)]

    def procesar_bloque(ii):
        iM = min(ii + BLOQUE, n)
        for kk in range(0, n, BLOQUE):
            for jj in range(0, n, BLOQUE):
                kM = min(kk + BLOQUE, n)
                jM = min(jj + BLOQUE, n)
                for i in range(ii, iM):
                    for k in range(kk, kM):
                        aik = A[i][k]  # reuso de registro
                        for j in range(jj, jM):
                            C[i][j] += aik * B[k][j]

    hilos = []
    for ii in range(0, n, BLOQUE):
        t = threading.Thread(target=procesar_bloque, args=(ii,))
        hilos.append(t)
        t.start()
    for t in hilos:
        t.join()
    return C
